### PR TITLE
CI: build CMake 3.20 to support LLVM 17

### PR DESCRIPTION
**Pull request by Kobzol**
LLVM 17 will require CMake at least 3.20, so we have to go back to building our own CMake on the Linux x64 dist builder.

r? @nikic

### CODE REVIEW

**Comment by Kobzol** [comment]
@rust-timer build 925c053abfb4df2511f5c69a1a21db5e3c09f87a

**Comment by Kobzol** [comment]
:grin: Looks like updating to LLVM 17 (as the host compiler) will be fun. But we don't need to deal with that now.

@rustbot ready

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+mkdir cmake-build
+cd cmake-build
+hide_output ../cmake-$CMAKE/configure --prefix=/rustroot
+hide_output make -j$(nproc)
+hide_output make install
+
+cd ..
+rm -rf cmake-build
+rm -rf cmake-$CMAKE
```
Comment (nikic) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh:17]: Shouldn't this be in src/ci/scripts? We're going to need this at least on dist-i686-linux and on all images that have Ubuntu < 22.04 (unless we update them).

Comment (Kobzol) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh:17]: The 32-bit runner uses the scripts from the x64 directory (https://github.com/rust-lang/rust/blob/eadf69a6c6edfe220fc5b1b659e46e271d75a3a1/src/ci/docker/host-x86_64/dist-i686-linux/Dockerfile#L53 :man_shrugging:). But good point that we should install CMake on more runners. Probably to all runners that were modified in https://github.com/rust-lang/rust/pull/95026?

Comment (nikic) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh:17]: Should be all runners that use either `ubuntu:18.04` or `ubuntu:20.04`.

**Comment by nikic** [comment]
Those results are certainly wild ... the largest regressions are on check builds, while I would have assumed that the host compiler shouldn't affect rustc itself much.

My first guess was that the new llvm-profdata can't handle the profiles generated by LLVM 16 and we lose PGO on rustc, but at least the log doesn't indicate something like that.

**Comment by Kobzol** [comment]
My guess was also the profiles, but as you said, logs look OK. BOLT could become worse, but such a large effect definitely shouldn't be BOLT alone. And it's not used in check builds anyway.

Debug/Opt builds could become worse if LLVM 17 optimized src/llvm-project worse, but why are checks builds bad, that's a mastery.

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh (hunk 1 above)
Comment (klensy) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh:17]: Shouldn't use https://github.com/rust-lang/rust/blob/master/src/ci/docker/scripts/cmake.sh, only move version out?

Comment (nikic) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-cmake.sh:17]: Yeah, we should bump the version in that script and use it.

**Comment by Kobzol** [comment]
I went through the builders modified in https://github.com/rust-lang/rust/pull/95026, most of them now use Ubuntu 22.04, which has new enough CMake (3.22). I updated the rest. I'm not sure which builders do actually build LLVM, so if I missed some, we'll probably see it when we bump LLVM to 17.

**Comment by nikic** [comment]
> I went through the builders modified in #95026, most of them now use Ubuntu 22.04, which has new enough CMake (3.22). I updated the rest. I'm not sure which builders do actually build LLVM, so if I missed some, we'll probably see it when we bump LLVM to 17.

Doesn't everything build LLVM on commits that update LLVM?

**Comment by Kobzol** [comment]
Good point. I went through the non-dist builders and added CMake where they use Ubuntu 20.04. I wonder if we should just upgrade them to 22.04..

**Comment by Kobzol** [comment]
Apparently dash couldn't handle the `cmake` script, so I switched everything to `bash`.

**Comment by Kobzol** [comment]
```
2023-07-16T10:06:12.7244668Z -- Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY OPENSSL_INCLUDE_DIR) 
2023-07-16T10:06:12.7245084Z CMake Error at Utilities/cmcurl/CMakeLists.txt:511 (message):
2023-07-16T10:06:12.7245405Z   Could not find OpenSSL.  Install an OpenSSL development package or
2023-07-16T10:06:12.7245795Z   configure CMake with -DCMAKE_USE_OPENSSL=OFF to build without OpenSSL.
2023-07-16T10:06:12.7245980Z 
2023-07-16T10:06:12.7245988Z 
2023-07-16T10:06:12.7246147Z -- Configuring incomplete, errors occurred!
2023-07-16T10:06:12.7246474Z See also "/cmake-build/CMakeFiles/CMakeOutput.log".
2023-07-16T10:06:12.7246799Z See also "/cmake-build/CMakeFiles/CMakeError.log".
2023-07-16T10:06:12.7247108Z ---------------------------------------------
2023-07-16T10:06:12.7247353Z Error when bootstrapping CMake:
2023-07-16T10:06:12.7247586Z Problem while running initial CMake
2023-07-16T10:06:12.7247871Z ---------------------------------------------
2023-07-16T10:06:12.9263238Z The command '/bin/sh -c /scripts/cmake.sh' returned a non-zero code: 1
```
Uhh, apparently CMake (or rather, `curlc`) actually requires OpenSSL now. Should I disable it (`-DCMAKE_USE_OPENSSL=OFF`) or install `libssl-dev` everywhere that it's missing? It's already installed in a lot of the builders, so I suppose that it wouldn't hurt to install it in the rest.

**Comment by Kobzol** [comment]
It was missing in only three runners, unless I missed something, so I added it to them.

**Comment by Kobzol** [comment]
How can updating CMake cause *linker errors*? What madness is this. Apparently `backtrace` isn't found now on `netbsd`. Maybe it could be caused simply by its Dockerfile updating, and e.g. `apt` packages changing versions from the last Docker rebuild?

**Comment by nikic** [comment]
Previously:
```
2023-07-16T19:05:57.5381369Z -- Looking for backtrace
2023-07-16T19:05:57.5668980Z -- Looking for backtrace - not found
2023-07-16T19:05:57.5675002Z -- Could NOT find Backtrace (missing: Backtrace_LIBRARY) 
```
Now:
```
2023-07-16T15:14:41.6919690Z -- Looking for backtrace
2023-07-16T15:14:41.7210383Z -- Looking for backtrace - not found
2023-07-16T15:14:41.7218608Z -- Found Backtrace: /x-tools/x86_64-unknown-netbsd/sysroot/usr/lib/libexecinfo.so
```

So it looks like cmake now detects backtrace support on NetBSD, but we fail to properly link against libexecinfo.

**Comment by nikic** [comment]
And I think the reason why we don't link it is that for cross-compilation we have to manually specify all the needed libraries in compiler/rustc_llvm/build.rs.

**Comment by Kobzol** [comment]
Is it "safe" to link to `libexecinfo` on this target, i.e. is there a chance that it will break something? Apparently before it worked even without linking to it.

**Comment by nikic** [comment]
I think it should be safe to link. Alternatively, we could also disable it via `-DCMAKE_DISABLE_FIND_PACKAGE_Backtrace=TRUE` or so.

**Comment by Kobzol** [comment]
Ok, I tried linking to it. I also added netbsd as a temporary PR CI job to test if it works.

**Comment by Kobzol** [comment]
Ok, looks like that has worked! (The failure is just being unable to upload artifacts to S3 from a PR CI run).

**Comment by Kobzol** [comment]
As much as it is suspicious in an iffy PR like this, `2023-07-17T12:06:44.8888610Z curl: (56) OpenSSL SSL_read: Connection reset by peer, errno 104` looks.. spurious?

@bors retry

**Comment by klensy** [comment]
Looking at fails in https://github.com/rust-lang/rust/pull/114048#issuecomment-1649586577
In most docker files cmake installed via apt, created symlink into rustroot, and later build 3.20 via cmake.sh. Should previous version be deleted?

https://github.com/rust-lang-ci/rust/actions/runs/5577768220/job/15104051656#step:24:1907

```
    Your CMake version is 3.17.5.  Starting with LLVM 17.0.0, the minimum
    version of CMake required to build LLVM will become 3.20.0, and using an
    older CMake will become an error.  Please upgrade your CMake to at least
    3.20.0 now to avoid issues in the future!
```

**Pull request merged**